lib/FrameManager.js

```diff
@@ -614,6 +621,49 @@ class Frame {
     }
   }
 
+  /**
+   * @param {string} xpath
+   * @param {!Object=} options
+   * @return {!Promise}
+   */
+  waitForXpath(xpath, options = {}) {
+    const timeout = options.timeout || 30000;
+    const waitForVisible = !!options.visible;
+    const waitForHidden = !!options.hidden;
+    const polling = waitForVisible || waitForHidden ? 'raf' : 'mutation';
+    return this.waitForFunction(predicate, {timeout, polling}, xpath, waitForVisible, waitForHidden);
+
+    /**
+     * @param {string} xpath
+     * @param {boolean} waitForVisible
+     * @param {boolean} waitForHidden
+     * @return {?Node|boolean}
+     */
+    function predicate(xpath, waitForVisible, waitForHidden) {
```

Comment:
This has a lot of code shared with waitForSelector. I think we should pop it out into its own function, and add an xpath/selector argument to it.

Reply:
removed some duplication.